tests/compiler/test_typing.py

```diff
@@ -103,31 +103,162 @@ def test_int_type_invalid(snippetcompiler):
     )
 
 
+def test_number_type(snippetcompiler):
+    snippetcompiler.setup_for_snippet(
+        """
+entity Test:
+    number i = 0
+end
+
+implement Test using std::none
+
+Test(i = 42)
+Test(i = -42)
+Test()
+Test(i = 42.0)
+Test(i = false)
+        """,
+    )
+    compiler.do_compile()
+
+
 def test_cast_to_number(snippetcompiler):
     snippetcompiler.setup_for_snippet(
         """
-x = number("42")
-x = 42
-y = number("42.0")
-y = 42.0
+a = number(21)
```

Comment:
Huh, this doesn't seem to correspond with my testing, see my other comment. Any clue how comes?

Reply:
indeed it seems this doesn't work:
```
    entity Test:
        float i = 0
    end

    entity Test2:
        int i = 0
    end
    
    implement Test using std::none
    implement Test2 using std::none
    
    val1 = Test(i = 42.0)
    val2 = Test2(i = 42)
    val1 = val2
```
So i'm a bit surprised, I'll look tomorrow

Reply:
I think this is fixed now

Reply:
Oh, it must be the `.0`: `42 == 42.0` even if `42 is not 42.0`

Reply:
So I'd use `42.1` or something for this test case.

Reply:
The problem is that you set up the failure scenario to raise a double set exception. Only, `42 == 42.0`, so `x = 42 x = 42.0` is perfectly valid (though admittedly a bit strange, and execution order would affect the actual type). Therefore even if the cast didn't work correctly (which it didn't) you would still get the correct answer. By making it `42.1`, when the bug is triggered, we get `42.1 == 42`, which is trivially wrong, therefore a more reliable assertion.